dask/utils.py

```diff
@@ -877,3 +877,21 @@ def partial_by_order(*args, **kwargs):
     for i, arg in other:
         args2.insert(i, arg)
     return function(*args2, **kwargs)
+
+
+def is_arraylike(x):
+    """ Is this object a numpy array or something similar?
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> x = np.ones(5)
+    >>> is_arraylike(x)
+    True
+    >>> is_arraylike(5)
+    False
+    >>> is_arraylike('cat')
+    False
+    """
+    return (hasattr(x, '__array_ufunc__') or
+            hasattr(x, 'shape') and x.shape and hasattr(x, 'dtype'))
```

Comment:
I've added a naive definition of `is_arraylike` .  I suspect that this will need to shift a bit.  Critical feedback welcome.

Reply:
What's the actual use case for duck-typing here? It might be simpler to stick with `isinstance(x, np.ndarray)`.

Rather than checking for `__array_ufunc__`, which could be implemented on non-arrays (e.g., `xarray.Dataset`), it would probably be safer to check for `__array__`, which indicates that a class can be converted into a NumPy array.

Reply:
One relevant use case is turning pandas dataframes into sparse matrices in machine learning workloads

Reply:
I've replaced `__array_ufunc__` with `__array__`

Reply:
> I've replaced __array_ufunc__ with __array__

Did you forget to push? I still see `__array_ufunc__` above.

Reply:
This is showing `__array__` for me now.

@shoyer did the sparse-arrays-for-machine-learning example suffice?  Do you have further concerns here?

Reply:
Yes, this looks fine to me now. Note that `scipy.sparse` doesn't define `__array__` though.

Reply:
Yeah, it's also not ndarray compatible.  Sparse arrays from http://sparse.pydata.org/en/latest/ would work though.  I also would not be surprised to see scipy.sparse style CSR/CSC ndarray classes built sometime in the next six months.

Reply:
Just to be clear, the `is_arraylike` isn't a particularly solid check, nor is it the accurate `will_work_perfectly_with_dask_array` function that we really want.  I think it's worth having something that mostly works in order to flush out bugs and use cases.  I think I'm usually less conservative when it comes to these things than you tend to be though.  I'm also happy to revert back to just supporting np.ndarrays in the short term if you prefer.